File: src/pkg/schemas/wiki.py
from datetime import datetime

from pydantic import BaseModel, Field


class WikiPageCreate(BaseModel):
    id: str | None = None
    title: str
    page_type: str = Field(default="topic", pattern=r"^(topic|entity|concept|project|comparison)$")
    summary: str | None = None
    content: str = ""
    domains: list[str] = []
    tags: list[str] = []
    derived_from_notes: list[str] = []
    derived_from_sources: list[str] = []
    open_questions: list[str] = []
    confidence_score: float | None = Field(default=None, ge=0, le=1)


class WikiPageUpdate(BaseModel):
    title: str | None = None
    page_type: str | None = Field(
        default=None,
        pattern=r"^(topic|entity|concept|project|comparison)$",
    )
    summary: str | None = None
    content: str | None = None
    domains: list[str] | None = None
    tags: list[str] | None = None
    derived_from_notes: list[str] | None = None
    derived_from_sources: list[str] | None = None
    open_questions: list[str] | None = None
    confidence_score: float | None = Field(default=None, ge=0, le=1)
    needs_recompile: bool | None = None
    stale_reason: str | None = None


class WikiPageRead(BaseModel):
    model_config = {"from_attributes": True}

    id: str
    title: str
    page_type: str
    summary: str | None = None
    content: str
    domains: list[str]
    tags: list[str]
    derived_from_notes: list[str]
    derived_from_sources: list[str]
    open_questions: list[str]
    confidence_score: float | None = None
    needs_recompile: bool
    stale_reason: str | None = None
    stale_triggered_at: datetime | None = None
    last_compiled_at: datetime | None = None
    created_at: datetime
    updated_at: datetime


class WikiPageList(BaseModel):
    items: list[WikiPageRead]
    total: int


class WikiCompileRequest(BaseModel):
    title: str
    page_type: str = Field(default="topic", pattern=r"^(topic|entity|concept|project|comparison)$")
    note_ids: list[str] = []
    source_ids: list[str] = []
    instructions: str | None = None


class WikiRecompileSuggestionRead(BaseModel):
    model_config = {"from_attributes": True}

    id: int
    user_id: str
    wiki_id: str
    wiki_title: str | None = None
    trigger_type: str
    trigger_id: str
    reason: str
    evidence_preview: str | None = None
    status: str
    metadata_: dict | None = None
    created_at: datetime
    updated_at: datetime
    reviewed_at: datetime | None = None
    applied_at: datetime | None = None
    reviewer_note: str | None = None


class WikiRecompileSuggestionList(BaseModel):
    items: list[WikiRecompileSuggestionRead]
    total: int


class WikiSuggestionStatusUpdate(BaseModel):
    status: str = Field(pattern=r"^(pending|accepted|rejected|dismissed|applied)$")
    reviewer_note: str | None = None


class WikiSuggestRequest(BaseModel):
    trigger_type: str = Field(pattern=r"^(source|note)$")
    trigger_id: str
    limit: int = Field(default=5, ge=1, le=20)


class WikiCloneDraftRequest(BaseModel):
    title: str | None = None


class WikiEvidenceRefRead(BaseModel):
    ref_type: str
    ref_id: str
    title: str
    excerpt: str | None = None

File: personal_knowledge_graph/web/src/lib/api/wiki.ts
import { request } from "./client";

export interface WikiPage {
  id: string;
  title: string;
  page_type: string;
  summary?: string | null;
  content: string;
  domains: string[];
  tags: string[];
  derived_from_notes: string[];
  derived_from_sources: string[];
  open_questions: string[];
  confidence_score?: number | null;
  needs_recompile: boolean;
  stale_reason?: string | null;
  stale_triggered_at?: string | null;
  last_compiled_at?: string | null;
  metadata_?: Record<string, unknown> | null;
  created_at: string;
  updated_at: string;
}

export interface WikiPageList {
  items: WikiPage[];
  total: number;
}

export interface WikiPageCreate {
  id?: string;
  title: string;
  page_type?: string;
  summary?: string | null;
  content?: string;
  domains?: string[];
  tags?: string[];
  derived_from_notes?: string[];
  derived_from_sources?: string[];
  open_questions?: string[];
  confidence_score?: number | null;
}

export interface WikiPageUpdate {
  title?: string;
  page_type?: string;
  summary?: string | null;
  content?: string;
  domains?: string[];
  tags?: string[];
  derived_from_notes?: string[];
  derived_from_sources?: string[];
  open_questions?: string[];
  confidence_score?: number | null;
  needs_recompile?: boolean;
  stale_reason?: string | null;
}

export interface WikiRecompileSuggestion {
  id: number;
  user_id: string;
  wiki_id: string;
  wiki_title?: string | null;
  trigger_type: string;
  trigger_id: string;
  reason: string;
  evidence_preview?: string | null;
  status: "pending" | "accepted" | "rejected" | "dismissed" | "applied" | string;
  metadata_?: Record<string, unknown> | null;
  created_at: string;
  updated_at: string;
  reviewed_at?: string | null;
  applied_at?: string | null;
  reviewer_note?: string | null;
}

export interface WikiCloneDraftRequest {
  title?: string;
}

export interface WikiRecompileSuggestionList {
  items: WikiRecompileSuggestion[];
  total: number;
}

export interface WikiSuggestRequest {
  trigger_type: "source" | "note";
  trigger_id: string;
  limit?: number;
}

export interface WikiEvidenceRef {
  ref_type: string;
  ref_id: string;
  title: string;
  excerpt?: string | null;
}

export interface ReferenceResolveInput {
  ref_type?: string;
  ref_id?: string;
  type?: string;
  id?: string;
  title?: string;
  excerpt?: string | null;
}

export interface ReferenceRead {
  ref_type: string;
  ref_id: string;
  title: string;
  subtitle?: string | null;
  href?: string | null;
  excerpt?: string | null;
  status?: string | null;
  metadata_?: Record<string, unknown> | null;
}

export const wikiApi = {
  list: (params?: {
    page_type?: string;
    domain?: string;
    tag?: string;
    needs_recompile?: boolean;
    limit?: number;
    offset?: number;
  }) => {
    const q = new URLSearchParams();
    if (params?.page_type) q.set("page_type", params.page_type);
    if (params?.domain) q.set("domain", params.domain);
    if (params?.tag) q.set("tag", params.tag);
    if (params?.needs_recompile !== undefined) q.set("needs_recompile", String(params.needs_recompile));
    if (params?.limit) q.set("limit", String(params.limit));
    if (params?.offset) q.set("offset", String(params.offset));
    return request<WikiPageList>(`/wiki?${q}`);
  },
  get: (id: string) => request<WikiPage>(`/wiki/${encodeURIComponent(id)}`),
  create: (body: WikiPageCreate) =>
    request<WikiPage>("/wiki", { method: "POST", body: JSON.stringify(body) }),
  update: (id: string, body: WikiPageUpdate) =>
    request<WikiPage>(`/wiki/${encodeURIComponent(id)}`, {
      method: "PATCH",
      body: JSON.stringify(body),
    }),
  cloneDraft: (id: string, body?: WikiCloneDraftRequest) =>
    request<WikiPage>(`/wiki/${encodeURIComponent(id)}/clone-draft`, {
      method: "POST",
      body: JSON.stringify(body ?? {}),
    }),
  delete: (id: string) =>
    request<void>(`/wiki/${encodeURIComponent(id)}`, { method: "DELETE" }),
  suggestions: (params?: { status?: string; wiki_id?: string; limit?: number; offset?: number }) => {
    const q = new URLSearchParams();
    if (params?.status !== undefined) q.set("status", params.status);
    if (params?.wiki_id) q.set("wiki_id", params.wiki_id);
    if (params?.limit) q.set("limit", String(params.limit));
    if (params?.offset) q.set("offset", String(params.offset));
    return request<WikiRecompileSuggestionList>(`/wiki/suggestions?${q}`);
  },
  suggest: (body: WikiSuggestRequest) =>
    request<WikiRecompileSuggestion[]>("/wiki/suggestions", {
      method: "POST",
      body: JSON.stringify(body),
    }),
  updateSuggestion: (id: number, status: "pending" | "accepted" | "rejected" | "dismissed" | "applied", reviewer_note?: string | null) =>
    request<WikiRecompileSuggestion>(`/wiki/suggestions/${id}`, {
      method: "PATCH",
      body: JSON.stringify({ status, reviewer_note }),
    }),
  resolveReferences: (refs: ReferenceResolveInput[]) =>
    request<{ items: ReferenceRead[] }>("/wiki/references/resolve", {
      method: "POST",
      body: JSON.stringify({ refs }),
    }),
};
